test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,229 @@
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void equalsInvariantSubVariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv1 - i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void equalsInvariantSubVariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv1 - i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_I, "2"})
+    public void equalsVariantSubInvariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_L, "1"})
+    public void equalsVariantSubInvariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void notEqualsAddInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 != i`
+            if (inv1 + i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void notEqualsAdtLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 != i`
+            if (inv1 + i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void notEqualsInvariantSubVariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 != i`
+            if (inv1 - i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void notEqualsInvariantSubVariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 != i`
+            if (inv1 - i != inv2) {
+                blackhole();
+            }
+        }
+    }
```

Comment:
Why could you not have random arguments here?

Reply:
[notEqualsInvariantSubVariantLong.txt](https://github.com/openjdk/jdk/files/14015703/notEqualsInvariantSubVariantLong.txt)

Attaching the IdealGraphVisualizer file (renamed to `.txt` cause GitHub wants that).

When we use `Argument.NUMBER_42`, we can avoid generating traps. When `i == 0` we enter the if statement, otherwise we exit.

If we use random numbers, we most likely never enter the if statement so we can generate a trap there. In this case, the trap block as an extra `SubL` for `inv1 - i`.

Reply:
As long as you have inv1 == inv2, your comparison will have a different result when i == 0. 

Can you randomize just one value and use them for both inv1 and inv2?

Reply:
@caojoshua I see. That is a sad limitation. You could now add a `@Run` statement, but that is overkill as well. I hope to improve the IR framework to make it a bit easier to pass better arguments soon. For now I think we could just leave it with constants, for simplicity.

Reply:
> Can you randomize just one value and use them for both inv1 and inv2?

Yeah we can just take a single argument `inv1`. I'd prefer having two arguments and leaving them constant.